core/utils/codel.h

```diff
@@ -0,0 +1,301 @@
+#ifndef BESS_UTILS_CODEL_H_
+#define BESS_UTILS_CODEL_H_
+
+#include <cstring>
+#include <iostream>
+#include <string>
+
+#include "../kmod/llring.h"
+#include "../mem_alloc.h"
+#include "../packet.h"
+#include "time.h"
+
+#define DEFAULT_TARGET .05  // default delay target for codel
+#define DEFAULT_WINDOW .1   // default window size for codel
+#define DEFAULT_SLOTS 4096  // default number of slots in the codel queue
+
+/*
+  Codel(Controlled Delay Management) is an Queue controller based on this
+  article http://queue.acm.org/detail.cfm?id=2209336
+
+  It provides an active queue management to help prevent bufferbloat by dropping
+  queue entries at an increasing rate if the delay in the queue is above the
+  target queue delay. The equation used to calculate drop intervals is based on TCP
+  throughput response to drop probability.
+*/
+struct Codel {
```

Comment:
I am a bit confused -- why is this in C style. A lot of the functions accept `codel*` as input which seems like reasonable to include in a class function.

@barath said the reason was for similarity to `llring`, I find this form of doing this to be a bit strange, and also worry it is hard to maintain. I would rather one just made a very shallow wrapper around `llring` as a class, in which case the compiler will eliminate most if not all of the indirection costs.

Mostly my concern is that the current style looks a bit strange -- `llring` is in C style due to compatibility with the kernel module, I am not sure we should add code in this style when not necessary.

I would prefer rewriting this as a class in line with the standard C++ style used elsewhere in Bess.